sda-commons-server-testing/src/main/java/org/sdase/commons/server/testing/GoldenFileAssertions.java

```diff
@@ -0,0 +1,87 @@
+
+/**
+ * Special assertions for {@link Path} objects to check if a file matches the expected contents and
+ * updates them if needed.
+ *
+ * <p>These assertions are helpful to check if certain files are stored in the repository (like
+ * OpenAPI or AsyncApi).
+ */
+public class GoldenFileAssertions extends AbstractAssert<GoldenFileAssertions, Path> {
+  /**
+   * Constructor
+   *
+   * @param actual the path to test
+   */
+  public GoldenFileAssertions(Path actual) {
+    super(actual, GoldenFileAssertions.class);
+  }
+
+  /**
+   * Creates a new instance of {@link GoldenFileAssertions}
+   *
+   * @param actual the path to test
+   * @return the created assertion object
+   */
+  public static GoldenFileAssertions assertThat(Path actual) {
+    return new GoldenFileAssertions(actual);
+  }
+  /**
+   * Verifies that the text content of the actual {@code Path} is <b>exactly</b> equal to the given
+   * one. If not, an {@link AssertionError} is thrown, but in contrast to {@link
+   * org.assertj.core.api.PathAssert#hasContent(String)} the file is updated with the expected value
+   * so the next assert succeeds.
+   *
+   * <p>Use this assertion if you want to conveniently store the latest copy of a file in your
+   * repository, and let the CI fail if an update has not been committed.
+   *
+   * <p>Examples:
+   *
+   * <pre><code class="java">
+   * Path xFile = Paths.get("openapi.yaml");
+   *
+   * String expected = ...; // call the service / start the generator
+   *
+   * GoldenFileAssertions.assertThat(xFile).hasContentOnDisk(expected);
+   * </code></pre>
+   *
+   * @param expected the expected text content to compare the actual {@code Path}'s content to.
+   * @return {@code this} assertion object.
+   * @throws NullPointerException if the given content is {@code null}.
+   * @throws UncheckedIOException if an I/O error occurs.
+   * @throws AssertionError if the actual {@code Path} is {@code null}.
+   * @throws AssertionError if the actual {@code Path} is not a {@link Files#isReadable(Path)
+   *     readable} file.
+   * @throws AssertionError if the content of the actual {@code Path} is not equal to the given
+   *     content.
+   */
+  public GoldenFileAssertions hasContentAndUpdateGolden(String expected) throws IOException {
+    // check if path is not null
+    isNotNull();
+
+    try {
+      // assert the file
+      String fileName = actual.getFileName().toString();
+      Assertions.assertThat(actual)
+          .as(
+              "The current %s file is not up-to-date. If this happens locally, "
+                  + "just run the test again. The %s file is updated automatically after running "
+                  + "this test. If this happens in the CI, make sure that you have committed the "
+                  + "latest %s file!",
+              fileName, fileName, fileName)
+          .hasContent(expected);
+    } finally {
+      // always update the file content
+      Files.write(actual, expected.getBytes());
```

Comment:
As we promote this assertion for API specs, should we provide a specific Yaml-Assertion and write the file only if it differs semantically (or does not exist for initial setup)?
We had many cases in the past where no semantic changes occur due to technical reasons in the past. That caused useless changes that appeared in reviews.

Reply:
We had several PRs in the past that fixed the non-semantic differences. At least from our experience they don't exist anymore.
In the base for this code we used Yaml based recursive assertions, but the output was really hard to understand. The `hasContent` assertion has a nice output that humans can understand.

Reply:
Well, I see such changes regularly in PRs.

Reply:
Then we should fix the root cause instead of patching it somewhere else

Reply:
I added an `assertThatYaml` variant that ignores key-order. But I would also prefer to fix the root-cause since I also seen a large number of PRs that update the file even-though there were no updates. And that makes the review harder.